### PR TITLE
B-enhancement/gh-538-import-leave-requests

### PR DESCRIPTION
**Pull request by mariobasic**
[#538] Introduce logic for importing(updating) requests from appogeeleave

 - Add classpath resource to application properties

 - Remove calculating allowances from RequestDbWriter will be done
    separately

 - Rename User and Request FiledSetMapper

 - Add another job for importing from csv file to RequestBatchJob

 - Add Data class for mapping from file to property

 - Add custom queried method for getting requests with days to
    RequestRepository

### CODE REVIEW

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    persistedRequests.forEach(
+        request -> mappedRequests.merge(request.getUser().getEmail(), List.of(request), (oldReq, newReq) -> {
+          List<Request> requests = new ArrayList<>(oldReq);
+          requests.add(request);
+          return requests;
+        }));
+  }
+
+  @Override
+  public Request process(AppogeeLeaveRequestImportData leaveImport) throws Exception {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: You can remove the `throws` declaration.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    List<Request> allUsersRequests = mappedRequests.get(userEmail);
+    List<Request> filteredRequestsMap = allUsersRequests.stream()
+                                                        .filter(requestsSubmittedOnThisDate(dateSubmitted))
+                                                        .filter(request -> isBetween(firstDay, lastDay, request))
+                                                        .filter(requestsWithSameAbsenceName(absenceName))
+                                                        .filter(request -> request.getReason().equals(reason))
+                                                        .distinct()
+                                                        .collect(Collectors.toList());
+
+    if (filteredRequestsMap.size() != 1) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Add empty line after this one, to improve readability.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+      if (filteredRequestsMap.size() == 0 && (status.equals("cancelled") || status.equals("rejected"))) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Instead of checking `size() == 0`, on collections, always use `isEmpty(), the code will be more expressive.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        return createRequestFromImport(leaveImport);
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Add empty line after this one, to improve readability.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+      } else if (filteredRequestsMap.size() == 2) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Add empty line after this one, to improve readability.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        if (isDurationBetweenRequestsDifferent(filteredRequestsMap)) {
+          removeRequestFromListOfUsersRequests(userEmail, allUsersRequests, filteredRequestsMap);
+          return updateRequest(filteredRequestsMap.get(0), leaveImport);
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Add empty line after this one, to improve readability.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+    return importedRequest;
+  }
+
+  private List<Day> createDaysFromImport(AppogeeLeaveRequestImportData leaveImport) {
+    List<Day> days = new ArrayList<>();
+    leaveImport.getStartDate()
+               .datesUntil(leaveImport.getEndDate().plusDays(1))
+               .filter(isWorkDay())
+               .forEach(date -> days.add(getDay(leaveImport, date)));
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Why not use map and collect, instead?
```
return leaveImport.getStartDate()
               .datesUntil(leaveImport.getEndDate().plusDays(1))
               .filter(isWorkDay())
               .map(date -> getDay(leaveImport, date))
               .collect(Collectors.toList());
```

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+    return days;
+  }
+
+  private Predicate<LocalDate> isWorkDay() {
+    return date -> !(DayOfWeek.SATURDAY.equals(date.getDayOfWeek()) || DayOfWeek.SUNDAY.equals(date.getDayOfWeek())
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Extract local variable to make the code more expressive:
```
boolean isWeekend = DayOfWeek.SATURDAY.equals(date.getDayOfWeek()) || DayOfWeek.SUNDAY.equals(date.getDayOfWeek();
boolean isPublicHoday = PUBLIC_HOLIDAY_DATES.contains(date) 
return !isWeekend && !isPublicHoliday;
```

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+   * @return duration for date.
+   */
+  private Duration getDuration(AppogeeLeaveRequestImportData leaveImport, LocalDate date) {
+    var startDate = leaveImport.getStartDate();
+    var endDate = leaveImport.getEndDate();
+    var calculatedRequestedDays = leaveImport.getStartDate()
+                                             .datesUntil(leaveImport.getEndDate().plusDays(1))
+                                             .filter(isWorkDay())
+                                             .count();
+    float calculatedRequestedHours = calculatedRequestedDays * FULL_DAY;
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Why do you use float here, when it is all integer values? It can lead to errors in comparison. If you used float because `calculatedRequestDays` is returned as `long`, you should also use `long` in variable definition.

See these articles:
[Correct way to compare floats or doubles in Java](https://howtodoinjava.com/java/basics/correctly-compare-float-double/)
[Comparing floating point](http://www.fredosaurus.com/notes-java/data/expressions/comparing-floatingpoint.html)

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+   * @param allUsersRequests list of all requests that given user has made.
+   * @param filteredRequestsMap map or users and theirs requests.
+   */
+  private void removeRequestFromListOfUsersRequests(String userEmail, List<Request> allUsersRequests,
+      List<Request> filteredRequestsMap) {
+    allUsersRequests.remove(filteredRequestsMap.get(0));
+    mappedRequests.put(userEmail, allUsersRequests);
+  }
+
+  private boolean isRequestCopyOfItSelf(List<Request> filteredRequestsMap) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: Capitalization is not correct, it should be `isRequestCopyOfItself()`. *ItSelf* -> *Itself*

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+  }
+
+  private boolean isDurationBetweenRequestsDifferent(List<Request> filteredRequestsMap) {
+    var firstRequestDuration = filteredRequestsMap.get(0).getDays().get(0).getDuration();
+    var secondRequestDuration = filteredRequestsMap.get(1).getDays().get(0).getDuration();
+
+    return firstRequestDuration != secondRequestDuration;
+  }
+
+  private Predicate<Request> requestsSubmittedOnThisDate(LocalDate dateSubmitted) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: `requestsSubmittedOnThisDate` -> `requestsSubmittedOnDate`

File: src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  }
+
+  @Override
+  public void setAsText(String text) throws IllegalArgumentException {
+    if (!StringUtils.isEmpty(text)) {
+
+      if (dateFormat.equals(LOCAL_DATE_FORMAT)) {
+        setValue(LocalDate.parse(text, formatter));
+      } else
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java]: Never omit brackets! Please add them around line 45.

File: src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+  @Override
+  public String getAsText() throws IllegalArgumentException {
+    Object date = getValue();
+    if (date != null) {
+
+      if (dateFormat.equals(LOCAL_DATE_FORMAT)) {
+        return formatter.format((LocalDate) date);
+      } else
+        return formatter.format(((LocalDateTime) date));
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java]: Never omit brackets! Please add them around this line.

File: src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  CustomPropertyEditorSupport(String dateFormat) {
+    super();
+    this.dateFormat = dateFormat;
+    this.formatter = DateTimeFormatter.ofPattern(dateFormat);
+  }
+
+  @Override
+  public String getAsText() throws IllegalArgumentException {
+    Object date = getValue();
+    if (date != null) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/CustomPropertyEditorSupport.java]: You could exit early and simplify the logic after this check if you invert the condition:
```
if (date == null) {
  return "";
}

if (dateFormat.equals(LOCAL_DATE_FORMAT) {
  return formatter.format((LocalDate) date);
} else {
  return formatter.format(((LocalDateTime) date));
}
```

File: src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java [hunk 15]
```diff
@@ -61,6 +76,19 @@ public Step calendarStep1(ItemReader<CalendarImportData> itemReader,
     //@formatter:om
   }
 
+  @Bean
+  public Step appogeeLeaveStep1(ItemReader<AppogeeLeaveRequestImportData> itemReader,
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java]: Why does this method need to be public and declared as a Bean? Same with `calendarStep1` and `step2` methods.

File: src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java [hunk 16]
```diff
@@ -99,4 +139,25 @@ public Step step2(Resource resource) {
     return defaultLineMapper;
   }
 
+  private LineMapper<AppogeeLeaveRequestImportData> appogeeLeaveLineMapper() {
+    DefaultLineMapper<AppogeeLeaveRequestImportData> defaultLineMapper = new DefaultLineMapper<>();
+    String[] fields = {"Name", "Email", "Parent Team", "Team", "EmployeeStatus", "PayrollNo", "DateSubmitted",
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java]: Extract this as a constant.

File: src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java [hunk 17]
```diff
@@ -90,7 +130,7 @@ public Step step2(Resource resource) {
     lineTokenizer.setNames("Event Title", "Event Description", "Reason", "Event Start", "Event End", "Date Created",
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java]: Extract the parameter array as a constant.

File: src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java [hunk 18]
```diff
@@ -22,16 +23,19 @@
 import org.springframework.core.io.Resource;
 
 @Configuration
-public class CalendarBatchJob {
+public class RequestBatchJob {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/RequestBatchJob.java]: Proper name would be `RequestBatchJobConfiguration`.

File: src/main/java/eu/execom/hawaii/configuration/UserBatchJob.java [hunk 19]
```diff
@@ -92,7 +92,7 @@ public Step userStep2() {
     lineTokenizer.setNames("FirstName", "LastName", "Email", "PayrollNo", "ContinuousStartDate", "StartDate", "EndDate",
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/configuration/UserBatchJob.java]: Proper name of this class would be `UserBatchJobConfiguration`.

Comment (stanislavgrujic) [reply src/main/java/eu/execom/hawaii/configuration/UserBatchJob.java]: Recheck this class for methods which are unnecessary `public` and which create useless beans.

File: src/main/resources/data.sql [hunk 20]
```diff
@@ -1,6 +1,6 @@
 LOCK TABLES `icon` WRITE;
 /*!40000 ALTER TABLE `icon` DISABLE KEYS */;
-INSERT INTO `icon` VALUES 
+insert into `icon` VALUES
```
Comment (stanislavgrujic) [inline src/main/resources/data.sql]: Why did you change case here? Now you have mix of upper and lower case for keywords (`insert`, `into`, `VALUES`). Please revert

File: src/main/resources/db/changelog/changelog-master.xml [hunk 21]
```diff
@@ -5,7 +5,7 @@
 
   <property name="initialId" value="initial_data" dbms="mysql"/>
   <property name="migrationId" value="migration" dbms="mysql"/>
-  <property name="collationId" value="collation" dbms="mysql"/>
+    <property name="collationId" value="collation" dbms="mysql"/>
```
Comment (stanislavgrujic) [inline src/main/resources/db/changelog/changelog-master.xml]: This line is not aligned with other lines. Please fix spacing.

File: src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+                                                       .stream()
+                                                       .filter(request -> request.getRequestStatus()
+                                                                                 .equals(RequestStatus.APPROVED))
+                                                       .collect(Collectors.toList());
+
+    persistedRequests.forEach(
+        request -> mappedRequests.merge(request.getUser().getEmail(), List.of(request), (oldReq, newReq) -> {
+          List<Request> requests = new ArrayList<>(oldReq);
+          requests.add(request);
+          return requests;
```
[the remark this answers is not in the document]
Comment (mariobasic) [reply src/main/java/eu/execom/hawaii/batch/AppogeeLeaveProcessor.java]: List `oldReq` is immutable so I didn't know any other way to do this but to create new list and pass 'oldReq' list through constructor and create a new one. With requested changes in comment bellow function will look like this: 
```
        persistedRequests.forEach(
            request -> mappedRequests.merge(request.getUser().getEmail(), List.of(request), (requests, newRequest) -> {
              List<Request> updatedRequests = new ArrayList<>(requests);
              updatedRequests.add(request);
              return updatedRequests;
            }));
```
Is there a better way to do this? Is variable naming OK?

File: src/main/resources/data.sql (hunk 20 above)
Comment (mariobasic) [reply src/main/resources/data.sql]: I couldn't remember doing this changes and started investigating what could have happened. All in all when you apply formatting (which I have set to be done automatically during commit) for some reason Intellij lowers caption only for `insert into` keyword.
